crates/compiler/derive/src/inspect.rs

```diff
@@ -0,0 +1,1096 @@
+                .collect::<Vec<(Lowercase, _)>>();
+            let fields = RecordFields::insert_into_subs(env.subs, flex_fields);
+            let record_var = synth_var(
+                env.subs,
+                Content::Structure(FlatType::Record(fields, Variable::EMPTY_RECORD)),
+            );
+
+            to_inspector_record(env, record_var, fields, def_symbol)
+        }
+        FlatInspectableKey::Tuple(arity) => {
+            // Generalized tuple var so we can reuse this impl between many tuples:
+            // if arity = n, this is (t1, ..., tn) for fresh t1, ..., tn.
+            let flex_elems = (0..arity)
+                .map(|idx| (idx as usize, env.subs.fresh_unnamed_flex_var()))
+                .collect::<Vec<_>>();
+            let elems = TupleElems::insert_into_subs(env.subs, flex_elems);
+            let tuple_var = synth_var(
+                env.subs,
+                Content::Structure(FlatType::Tuple(elems, Variable::EMPTY_TUPLE)),
+            );
+
+            to_inspector_tuple(env, tuple_var, elems, def_symbol)
+        }
+        FlatInspectableKey::TagUnion(tags) => {
+            // Generalized tag union var so we can reuse this impl between many unions:
+            // if tags = [ A arity=2, B arity=1 ], this is [ A t1 t2, B t3 ] for fresh t1, t2, t3
+            let flex_tag_labels = tags
+                .into_iter()
+                .map(|(label, arity)| {
+                    let variables_slice =
+                        VariableSubsSlice::reserve_into_subs(env.subs, arity.into());
+                    for var_index in variables_slice {
+                        env.subs[var_index] = env.subs.fresh_unnamed_flex_var();
+                    }
+                    (label, variables_slice)
+                })
+                .collect::<Vec<_>>();
+            let union_tags = UnionTags::insert_slices_into_subs(env.subs, flex_tag_labels);
+            let tag_union_var = synth_var(
+                env.subs,
+                Content::Structure(FlatType::TagUnion(
+                    union_tags,
+                    TagExt::Any(Variable::EMPTY_TAG_UNION),
+                )),
+            );
+
+            to_inspector_tag_union(env, tag_union_var, union_tags, def_symbol)
+        }
+        FlatInspectableKey::Function(_arity) => {
+            // Desired output: \x, y, z -> ... ===> "<function>"
+
+            todo!();
+        }
+        FlatInspectableKey::Opaque => todo!(),
+        FlatInspectableKey::Error => todo!(),
+    };
+
+    let specialization_lambda_sets =
+        env.get_specialization_lambda_sets(body_type, Symbol::INSPECT_TO_INSPECTOR);
+
+    DerivedBody {
+        body,
+        body_type,
+        specialization_lambda_sets,
+    }
+}
+
+fn to_inspector_list(env: &mut Env<'_>, fn_name: Symbol) -> (Expr, Variable) {
+    // Build \lst -> list, List.walk, (\elem -> Inspect.toInspector elem)
+    //
+    // TODO eta reduce this baby       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Comment:
We can get rid of this comment. We now know that we can't eta-reduce this.

Reply:
Done